src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisState.cs

```diff
@@ -771,52 +783,60 @@ public void MarkDeclarationsComplete(ISymbol symbol, IEnumerable<DiagnosticAnaly
-        public void MarkSyntaxAnalysisComplete(SourceOrAdditionalFile file, DiagnosticAnalyzer analyzer)
+        public bool TryMarkSyntaxAnalysisComplete(SourceOrAdditionalFile file, DiagnosticAnalyzer analyzer)
         {
-            GetAnalyzerState(analyzer).MarkSyntaxAnalysisComplete(file);
+            return GetAnalyzerState(analyzer).TryMarkSyntaxAnalysisComplete(file);
         }
 
         /// <summary>
-        /// Marks the given file as fully syntactically analyzed for the given analyzers.
+        /// Tries to mark the given file as fully syntactically analyzed for the given analyzers.
         /// </summary>
-        public void MarkSyntaxAnalysisComplete(SourceOrAdditionalFile file, IEnumerable<DiagnosticAnalyzer> analyzers)
+        public bool TryMarkSyntaxAnalysisComplete(SourceOrAdditionalFile file, IEnumerable<DiagnosticAnalyzer> analyzers)
         {
+            var success = true;
             foreach (var analyzer in analyzers)
             {
-                GetAnalyzerState(analyzer).MarkSyntaxAnalysisComplete(file);
+                if (!GetAnalyzerState(analyzer).TryMarkSyntaxAnalysisComplete(file))
+                    success = false;
             }
+
+            return success;
         }
 
         /// <summary>
-        /// Marks the given file as fully syntactically analyzed for the unprocessed analyzers in the given analysisScope.
+        /// Tries to mark the given file as fully syntactically analyzed for the unprocessed analyzers in the given analysisScope.
         /// </summary>
-        public void MarkSyntaxAnalysisCompleteForUnprocessedAnalyzers(
+        public bool TryMarkSyntaxAnalysisCompleteForUnprocessedAnalyzers(
             SourceOrAdditionalFile file,
             AnalysisScope analysisScope,
             HashSet<DiagnosticAnalyzer> processedAnalyzers)
-            => MarkAnalysisCompleteForUnprocessedAnalyzers(analysisScope, processedAnalyzers, MarkSyntaxAnalysisComplete, file);
+            => TryMarkAnalysisCompleteForUnprocessedAnalyzers(analysisScope, processedAnalyzers, TryMarkSyntaxAnalysisComplete, file);
 
-        private static void MarkAnalysisCompleteForUnprocessedAnalyzers<T>(
+        private static bool TryMarkAnalysisCompleteForUnprocessedAnalyzers<T>(
             AnalysisScope analysisScope,
             HashSet<DiagnosticAnalyzer> processedAnalyzers,
-            Action<T, DiagnosticAnalyzer> markComplete,
+            Func<T, DiagnosticAnalyzer, bool> tryMarkComplete,
             T arg)
         {
             Debug.Assert(processedAnalyzers.All(analysisScope.Contains));
             if (analysisScope.Analyzers.Length == processedAnalyzers.Count)
             {
-                return;
+                return true;
             }
 
+            var success = true;
             foreach (var analyzer in analysisScope.Analyzers)
             {
-                if (!processedAnalyzers.Contains(analyzer))
+                if (!processedAnalyzers.Contains(analyzer) &&
```

Comment:
>processedAnalyzers [](http://example.com/codeflow?start=21&length=18)

Here and in several other places we are using information on the side that tracks whether all things are marked as complete and is used to make various shortcuts. When we adjust that side information, do we take into account the fact that some of the states were not complete?

Reply:
Yes, this code path is only meant to handle analyzers that were completely unprocessed or skipped - for example `analysisScope` requested might only contain subset of analyzers, and rest can be skipped: https://sourceroslyn.io/#Microsoft.CodeAnalysis/DiagnosticAnalyzer/AnalyzerDriver.cs,747